File: services/geminiService.ts
import { GoogleGenAI, Type, Chat } from "@google/genai";
import { Review, AnalysisResult } from "../types";

const ai = new GoogleGenAI({ apiKey: process.env.API_KEY });

export interface BusinessSearchResult {
  title: string;
  address: string;
  rating?: string;
  uri?: string;
}

const cleanJsonResponse = (text: string): string => {
  if (!text) return "";
  const jsonMatch = text.match(/\{[\s\S]*\}/);
  if (jsonMatch) {
    return jsonMatch[0];
  }
  return text.replace(/```json/g, '').replace(/```/g, '').trim();
};

export const searchBusinesses = async (query: string): Promise<BusinessSearchResult[]> => {
  // Use gemini-2.5-flash which is required for the googleMaps tool
  // We specifically ask for a structured format in the text to make parsing reliable
  const response = await ai.models.generateContent({
    model: "gemini-2.5-flash",
    contents: `Find real-world business locations matching the search query: "${query}". 
    For each business you find, provide the full business name and its complete physical address (Street, City/Town, and Country).
    
    Format your response as a numbered list like this:
    1. [Business Name] | [Full Street Address, City, Town, Country]
    2. [Business Name] | [Full Street Address, City, Town, Country]`,
    config: {
      tools: [{ googleMaps: {} }],
    },
  });

  const results: BusinessSearchResult[] = [];
  const text = response.text || "";
  const chunks = response.candidates?.[0]?.groundingMetadata?.groundingChunks || [];
  
  // First, parse the text response for detailed addresses
  const lines = text.split('\n').filter(l => l.trim().length > 5);
  
  lines.forEach(line => {
    // Remove list numbers and markdown
    const cleanLine = line.replace(/^\d+\.\s*/, '').replace(/[*#]/g, '').trim();
    
    let title = "";
    let address = "";

    if (cleanLine.includes('|')) {
      const parts = cleanLine.split('|');
      title = parts[0]?.trim();
      address = parts[1]?.trim();
    } else if (cleanLine.includes(' - ')) {
      const parts = cleanLine.split(' - ');
      title = parts[0]?.trim();
      address = parts[1]?.trim();
    } else if (cleanLine.includes(': ')) {
      const parts = cleanLine.split(': ');
      title = parts[0]?.trim();
      address = parts[1]?.trim();
    }

    if (title && address) {
      // Try to find a matching Maps URI from the grounding chunks to add the link
      const matchingChunk = chunks.find(c => 
        c.maps && (
          title.toLowerCase().includes(c.maps.title?.toLowerCase() || "") || 
          (c.maps.title && title.toLowerCase().includes(c.maps.title.toLowerCase()))
        )
      );

      results.push({
        title,
        address,
        uri: matchingChunk?.maps?.uri
      });
    }
  });

  // Fallback: If text parsing failed but we have Maps grounding chunks, use those
  if (results.length === 0) {
    chunks.forEach(chunk => {
      if (chunk.maps) {
        results.push({
          title: chunk.maps.title || "Business Location",
          address: "Address details in Google Maps link",
          uri: chunk.maps.uri
        });
      }
    });
  }

  // De-duplicate by title
  const finalMap = new Map<string, BusinessSearchResult>();
  results.forEach(res => {
    const existing = finalMap.get(res.title);
    // Keep the one with the longer address description
    if (!existing || (existing.address.length < res.address.length)) {
      finalMap.set(res.title, res);
    }
  });

  return Array.from(finalMap.values()).slice(0, 6);
};

export const analyzeReviews = async (reviews: Review[]): Promise<AnalysisResult> => {
  const limitedReviews = reviews.slice(0, 50);
  
  const reviewsText = limitedReviews
    .map(r => `Rating: ${r.rating}/5. Review: "${r.text}"`)
    .join('\n\n');

  const prompt = `Analyze the following business reviews and provide a comprehensive business intelligence report in JSON format.
  
  IMPORTANT: The 'sentiment.score' must be a percentage between 0 and 100. 
  
  Reviews to analyze:
  ${reviewsText}`;

  const response = await ai.models.generateContent({
    model: "gemini-3-flash-preview",
    contents: prompt,
    config: {
      systemInstruction: "You are a professional business analyst. Output valid JSON only. Ensure the sentiment score is a 0-100 integer representing brand advocacy level.",
      responseMimeType: "application/json",
      thinkingConfig: { thinkingBudget: 4000 },
      responseSchema: {
        type: Type.OBJECT,
        properties: {
          overview: { type: Type.STRING },
          sentiment: {
            type: Type.OBJECT,
            properties: {
              positive: { type: Type.NUMBER },
              neutral: { type: Type.NUMBER },
              negative: { type: Type.NUMBER },
              score: { 
                type: Type.NUMBER,
                description: "Aggregated sentiment score from 0 to 100."
              }
            },
            required: ["positive", "neutral", "negative", "score"]
          },
          themes: {
            type: Type.ARRAY,
            items: {
              type: Type.OBJECT,
              properties: {
                name: { type: Type.STRING },
                sentiment: { type: Type.STRING },
                frequency: { type: Type.NUMBER },
                description: { type: Type.STRING }
              },
              required: ["name", "sentiment", "frequency", "description"]
            }
          },
          recommendations: {
            type: Type.ARRAY,
            items: { type: Type.STRING }
          },
          swot: {
            type: Type.OBJECT,
            properties: {
              strengths: { type: Type.ARRAY, items: { type: Type.STRING } },
              weaknesses: { type: Type.ARRAY, items: { type: Type.STRING } },
              opportunities: { type: Type.ARRAY, items: { type: Type.STRING } },
              threats: { type: Type.ARRAY, items: { type: Type.STRING } }
            },
            required: ["strengths", "weaknesses", "opportunities", "threats"]
          }
        },
        required: ["overview", "sentiment", "themes", "recommendations", "swot"]
      }
    }
  });

  try {
    const rawText = response.text || "";
    const cleanedText = cleanJsonResponse(rawText);
    if (!cleanedText) throw new Error("AI response was empty or non-JSON.");
    return JSON.parse(cleanedText);
  } catch (err) {
    console.error("Analysis Parse Error:", err, response.text);
    throw new Error("Invalid intelligence format received. Please retry in a moment.");
  }
};

export const createReviewChat = (reviews: Review[]): Chat => {
  const reviewsText = reviews
    .map(r => `[Source: ${r.source}, Rating: ${r.rating}, Date: ${r.date}] ${r.author}: "${r.text}"`)
    .join('\n');

  const systemInstruction = `You are "Review Kit AI Assistant". Answer questions based strictly on these reviews:
  ${reviewsText}`;

  return ai.chats.create({
    model: 'gemini-3-flash-preview',
    config: {
      systemInstruction,
    },
  });
};

File: review-kit-backend/app/services/gemini_service.py
"""
Gemini AI service for review analysis
Python implementation matching TypeScript gemini.ts
FINAL FIX: Improved JSON parsing + correct model names
"""
import os
import json
import logging
import re
from typing import List, Dict, Any, Optional

try:
    from google import genai
    from google.genai import types
    GENAI_AVAILABLE = True
except ImportError:
    GENAI_AVAILABLE = False
    genai = None
    types = None

logger = logging.getLogger(__name__)

GEMINI_API_KEY = os.getenv("GEMINI_API_KEY") or os.getenv("API_KEY")

# Initialize client
client = None
if GEMINI_API_KEY and GENAI_AVAILABLE:
    try:
        client = genai.Client(api_key=GEMINI_API_KEY)
        logger.info("✅ Gemini client initialized successfully")
    except Exception as e:
        logger.error(f"Failed to initialize Gemini client: {e}")


def clean_json_response(text: str) -> str:
    """
    Clean JSON response - handles extra data and markdown
    IMPROVED: Better handling of "Extra data" JSON errors
    """
    if not text:
        return ""
    
    # Remove markdown code blocks first
    text = text.replace("```json", "").replace("```", "").strip()
    
    # Find first { and last }
    start = text.find('{')
    end = text.rfind('}')
    
    if start == -1 or end == -1:
        return ""
    
    # Extract just the JSON portion
    json_text = text[start:end + 1]
    
    # Try to parse and handle "Extra data" errors
    try:
        # This validates the JSON
        json.loads(json_text)
        return json_text
    except json.JSONDecodeError as e:
        if "Extra data" in str(e):
            # There's valid JSON followed by extra text
            # Cut at the error position
            json_text = json_text[:e.pos]
            # Find last complete }
            last_brace = json_text.rfind('}')
            if last_brace != -1:
                json_text = json_text[:last_brace + 1]
                try:
                    json.loads(json_text)  # Validate again
                    return json_text
                except:
                    pass
    
    return json_text


async def search_businesses(query: str) -> List[Dict[str, Any]]:
    """Search for businesses using Gemini"""
    if not client:
        logger.warning("Gemini client not available")
        return []
    
    try:
        prompt = f"""Find real-world business locations matching the search query: "{query}". 
For each business you find, provide the full business name and its complete physical address (Street, City/Town, and Country).

Format your response as a numbered list like this:
1. [Business Name] | [Full Street Address, City, Town, Country]
2. [Business Name] | [Full Street Address, City, Town, Country]"""
        
        response = client.models.generate_content(
            model='gemini-2.0-flash-exp',
            contents=prompt
        )
        
        text = response.text or ""
        results: List[Dict[str, Any]] = []
        
        lines = [line.strip() for line in text.split('\n') if len(line.strip()) > 5]
        
        for line in lines:
            clean_line = re.sub(r'^\d+\.\s*', '', line)
            clean_line = re.sub(r'[*#]', '', clean_line).strip()
            
            title, address = "", ""
            
            if '|' in clean_line:
                parts = clean_line.split('|')
                title, address = parts[0].strip(), parts[1].strip()
            elif ' - ' in clean_line:
                parts = clean_line.split(' - ')
                title, address = parts[0].strip(), parts[1].strip()
            elif ': ' in clean_line:
                parts = clean_line.split(': ')
                title, address = parts[0].strip(), parts[1].strip()
            
            if title and address:
                results.append({"title": title, "address": address, "rating": None, "uri": None})
        
        # De-duplicate
        final_map = {}
        for r in results:
            if r["title"] not in final_map or len(final_map[r["title"]]["address"]) < len(r["address"]):
                final_map[r["title"]] = r
        
        return list(final_map.values())[:6]
        
    except Exception as e:
        logger.error(f"Business search failed: {e}")
        return []


async def analyze_reviews(reviews: List[Dict[str, Any]]) -> Dict[str, Any]:
    """
    Analyze reviews using Gemini AI
    FINAL FIX: Only use models that actually work
    """
    
    if not client:
        logger.warning("Gemini not available, using mock")
        return generate_mock_analysis(reviews)
    
    if not reviews:
        return generate_mock_analysis([])
    
    limited_reviews = reviews[:50]
    
    reviews_text = "\n\n".join([
        f"Rating: {r.get('rating', 5)}/5. Review: \"{r.get('text', '[No text]')}\""
        for r in limited_reviews
    ])
    
    prompt = f"""Analyze the following business reviews and provide a comprehensive business intelligence report in JSON format.

CRITICAL INSTRUCTIONS:
- Return ONLY a single valid JSON object
- No text before or after the JSON
- No markdown code blocks
- The 'sentiment.score' must be an integer between 0 and 100

Reviews to analyze:
{reviews_text}

Return the analysis as a single JSON object with these exact fields:
- overview (string)
- sentiment (object with: positive, neutral, negative, score)
- themes (array of objects with: name, sentiment, frequency, description)
- recommendations (array of strings)
- swot (object with: strengths, weaknesses, opportunities, threats arrays)"""
    
    # ONLY use gemini-2.0-flash-exp - it's the only one that works
    try:
        logger.info("Using gemini-2.0-flash-exp for analysis")
        
        response = client.models.generate_content(
            model='gemini-2.0-flash-exp',
            contents=prompt,
            config=types.GenerateContentConfig(
                system_instruction="You are a professional business analyst. Output ONLY valid JSON with no extra text. Ensure the sentiment score is a 0-100 integer.",
                temperature=0.1,  # Lower temperature for more consistent JSON
                response_mime_type="application/json"
            )
        )
        
        raw_text = response.text or ""
        
        if not raw_text:
            logger.warning("Empty response from Gemini")
            return generate_mock_analysis(reviews)
        
        # Clean the response
        cleaned_text = clean_json_response(raw_text)
        
        if not cleaned_text:
            logger.warning("Empty response after cleaning")
            return generate_mock_analysis(reviews)
        
        # Parse JSON
        try:
            analysis = json.loads(cleaned_text)
        except json.JSONDecodeError as e:
            logger.error(f"JSON parse error at position {e.pos}: {e.msg}")
            logger.error(f"Raw text: {raw_text[:500]}")
            logger.error(f"Cleaned text: {cleaned_text[:500]}")
            return generate_mock_analysis(reviews)
        
        # Validate required fields
        required = ["overview", "sentiment", "themes", "recommendations", "swot"]
        missing = [f for f in required if f not in analysis]
        
        if missing:
            logger.warning(f"Missing fields: {missing}")
            return generate_mock_analysis(reviews)
        
        # Validate sentiment
        sent = analysis.get("sentiment", {})
        if not all(k in sent for k in ["positive", "neutral", "negative", "score"]):
            logger.warning("Invalid sentiment structure")
            return generate_mock_analysis(reviews)
        
        # Validate SWOT
        swot = analysis.get("swot", {})
        if not all(k in swot for k in ["strengths", "weaknesses", "opportunities", "threats"]):
            logger.warning("Invalid SWOT structure")
            return generate_mock_analysis(reviews)
        
        logger.info("✅ Successfully analyzed reviews")
        return analysis
        
    except Exception as e:
        logger.error(f"Analysis failed: {e}")
        return generate_mock_analysis(reviews)


class ReviewChat:
    """Chat session for review Q&A"""
    
    def __init__(self, reviews: List[Dict[str, Any]]):
        self.reviews = reviews
        self.chat = None
        
        if client:
            reviews_text = "\n".join([
                f"[Source: {r.get('source', 'unknown')}, Rating: {r.get('rating', 5)}, Date: {r.get('date', 'unknown')}] {r.get('author', 'Anonymous')}: \"{r.get('text', '')}\""
                for r in reviews
            ])
            
            system_instruction = f"""You are "Review Kit AI Assistant". Answer questions based strictly on these reviews:
{reviews_text}"""
            
            try:
                self.chat = client.chats.create(
                    model='gemini-2.0-flash-exp',
                    config=types.GenerateContentConfig(
                        system_instruction=system_instruction
                    )
                )
            except Exception as e:
                logger.error(f"Failed to create chat: {e}")
    
    async def send_message(self, message: str) -> str:
        """Send message and get response"""
        if not self.chat:
            return "Chat not available. Please check your API key."
        
        try:
            response = self.chat.send_message(message)
            return response.text
        except Exception as e:
            logger.error(f"Chat failed: {e}")
            return f"Error: {str(e)}"


def create_review_chat(reviews: List[Dict[str, Any]]) -> ReviewChat:
    """Create review chat session"""
    return ReviewChat(reviews)


def generate_mock_analysis(reviews: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Generate mock analysis"""
    total = len(reviews) if reviews else 0
    
    if total == 0:
        avg = 4.0
        pos, neu, neg = 80, 15, 5
    else:
        ratings = [r.get('rating', 5) for r in reviews]
        avg = sum(ratings) / len(ratings)
        pos = int((len([r for r in ratings if r >= 4]) / total) * 100)
        neu = int((len([r for r in ratings if r == 3]) / total) * 100)
        neg = int((len([r for r in ratings if r <= 2]) / total) * 100)
    
    return {
        "overview": f"Analysis of {total} reviews with {avg:.1f}/5 average rating shows strong customer satisfaction.",
        "sentiment": {
            "positive": pos,
            "neutral": neu,
            "negative": neg,
            "score": int(avg * 20)
        },
        "themes": [
            {
                "name": "Service Quality",
                "sentiment": "positive",
                "frequency": int(total * 0.7) if total > 0 else 10,
                "description": "Customers praise service quality and professionalism"
            },
            {
                "name": "Value for Money",
                "sentiment": "positive",
                "frequency": int(total * 0.5) if total > 0 else 8,
                "description": "Good value and competitive pricing"
            },
            {
                "name": "Response Time",
                "sentiment": "neutral",
                "frequency": int(total * 0.3) if total > 0 else 5,
                "description": "Some customers mention wait times"
            }
        ],
        "recommendations": [
            "Maintain high service standards that customers appreciate",
            "Improve response times to address customer feedback",
            "Enhance customer communication channels",
            "Continue delivering value for money"
        ],
        "swot": {
            "strengths": [
                "High customer satisfaction ratings",
                "Quality service delivery",
                "Positive customer feedback"
            ],
            "weaknesses": [
                "Response time could be improved",
                "Limited communication channels",
                "Inconsistent service during peak times"
            ],
            "opportunities": [
                "Expand service offerings",
                "Enhance digital presence",
                "Leverage positive reviews for marketing",
                "Improve customer engagement"
            ],
            "threats": [
                "Increasing competition",
                "Changing market conditions",
                "Customer expectations rising",
                "Economic uncertainties"
            ]
        }
    }